File: c++/30_Substring_with_Concatenation_of_All_Words.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>

using namespace std;

/*
For example, given:
"barfoothefoobarman"  ["foo", "bar"] return the indices: [0,9].
"barfoo-foo bar the-foobarman"  ["bar","foo","the"]   [6,9,12]
"wordgoodgoodgoodbestword", ["word","good","best","good"] [8]
*/

class Solution {
public:
    vector<int> findSubstring(string s, vector<string>& words) {
    	int strLen = s.size();
    	int wordLen = words[0].size();
    	int wordNum = words.size();

    	vector<int> res;
    	map<string, int> dict;
    	for (int i = 0; i < wordNum; i++)
    		dict[words[i]]++;
    	/*
    	map<string, int>::iterator it;
    	for (it = dict.begin(); it != dict.end(); it++) {
    		cout << it->second << endl;
    	}*/

    	if (strLen < wordNum * wordLen)
    		return res;
    	for (int i = 0; i < strLen - wordNum * wordLen + 1; i++) {
    		string subStr = s.substr(i, wordNum * wordLen);
    		if (match(i, wordNum, wordLen, subStr, dict, words))
    			res.push_back(i);
    	}

    	return res;
        
    }
    bool match(int start, int wordNum, int wordLen, string subString, map<string, int> &dict, vector<string> &words) {
    	map<string, int> cnt;
    	for (int i = 0; i < wordNum; i++) {   		
    		string subSubStr = subString.substr(i * wordLen, wordLen);
    		if (dict.count(subSubStr) > 0) {
    			cnt[subSubStr]++;
    			if (start == 0)
    				cout << subSubStr << " " << cnt[subSubStr] << " " <<  dict[subSubStr] << endl;
    		}
    		else
    			return 0;
    		if (cnt[subSubStr] > dict[subSubStr])
    			return 0;
    	}
    	return 1;
    }
};




int main() {
	Solution ans;
	string s = "barfoofoobarthefoobarman";
	vector<string> words;
	words.push_back("foo");
	words.push_back("bar");
	words.push_back("the");
	vector<int> res = ans.findSubstring(s, words);
	for (int i = 0; i < res.size(); i++)
		cout << res[i] << " ";
	cout << endl;



}

File: python/30_Substring_with_Concatenation_of_All_Words.py
#!/usr/bin/python
import sys
import os

'''
For example, given:
"barfoothefoobarman"  ["foo", "bar"] return the indices: [0,9].
"barfoo-foo bar the-foobarman"  ["bar","foo","the"]   [6,9,12]
"wordgoodgoodgoodbestword", ["word","good","best","good"] [8]
'''

'''
Bad question, ambiguous target, too different kinds of return values
'''


class Solution(object):
    def findSubstring(self, s, words):
    	if len(words) == 0:
    		return None
    	n = len(s)
    	p = len(words)
    	m = len(words[0])
    	word = {}
    	for i, element in enumerate(words):
    		word[element] = word.setdefault(element, 0) + 1

    	res, order, last = [], "", 0
    	for i in range(n - m + 1):
    		for j, element in enumerate(word): # enumerate foo, bar...
    			#print i, element
    			cnt = 0
    			for k in range(m): # f - o - o
    				if s[i + k] != element[k]:
    					break
    				cnt += 1
    			if cnt == m and ((i != 0 and i - last >= m) or i == 0):
    				#print i, last, i - last, "m = ", m, last
    				res.append(i)
    				order += str(j)
    				last = i

    	q = len(res)
    	#print "q = ", q
    	if q < p:
    		return None
    	cmp = ""
    	for i, element in enumerate(word):
    		for j in range(word[element]):
    			cmp += str(i)
    	#print cmp, p, order[0:2]
    	for i in range(q - p + 1):
			cat = "".join(sorted(order[i:(p + i)]))
			if cat == cmp:
				out = []
				for k in range(p):
					out.append(res[i + k])
				return out		
    	return None        


def main():
    s = Solution()
    print s.findSubstring("barfoothefoobarman", ["bar", "foo"])
    print s.findSubstring("barfoofoobarthefoobarman", ["bar","foo","the"])
    print s.findSubstring("wordgoodgoodgoodbestword", ["word","good","best","good"])
    print s.findSubstring("aaa", ["aa","aa"])
    print s.findSubstring("aaaaaa", ["aaa","aaa"])

if __name__ == "__main__":
    main()
